File: sdks/python/src/bugbarn/breadcrumbs.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional
from collections import deque
import datetime

MAX_BREADCRUMBS = 100


@dataclass
class Breadcrumb:
    category: str
    message: str
    timestamp: str = field(default_factory=lambda: datetime.datetime.now(datetime.timezone.utc).isoformat())
    level: Optional[str] = None
    data: Optional[Dict[str, Any]] = None

    def to_dict(self) -> dict:
        d: dict = {"timestamp": self.timestamp, "category": self.category, "message": self.message}
        if self.level:
            d["level"] = self.level
        if self.data:
            d["data"] = self.data
        return d


_buffer: deque[Breadcrumb] = deque(maxlen=MAX_BREADCRUMBS)


def add_breadcrumb(category: str, message: str, level: Optional[str] = None, data: Optional[Dict[str, Any]] = None) -> None:
    _buffer.append(Breadcrumb(category=category, message=message, level=level, data=data))


def get_breadcrumbs() -> List[Breadcrumb]:
    return list(_buffer)


def clear_breadcrumbs() -> None:
    _buffer.clear()

File: sdks/typescript/src/breadcrumbs.ts
export interface Breadcrumb {
  timestamp: string; // ISO 8601
  category: string; // "console" | "http" | "navigation" | "manual"
  message: string;
  level?: string; // "log" | "warn" | "error" | "info" | "debug"
  data?: Record<string, unknown>;
}

const MAX_BREADCRUMBS = 100;
const buffer: Breadcrumb[] = [];

export function addBreadcrumb(crumb: Breadcrumb): void {
  buffer.push(crumb);
  if (buffer.length > MAX_BREADCRUMBS) {
    buffer.shift();
  }
}

export function getBreadcrumbs(): Breadcrumb[] {
  return [...buffer];
}

export function clearBreadcrumbs(): void {
  buffer.length = 0;
}
